Move `short_to_chan_info` into standalone lock

ViktorTigerstrom opened the pull request:
Based on #1638 and a prerequisite for #1507.

As the `channel_state` (`ChannelHolder`) struct will be removed, this commit moves the `short_to_chan_info` map from that lock into a separate lock, and cleans up a few macros that can be cleaned up due to moving it into a separate lock.

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1379,44 +1386,44 @@ macro_rules! handle_monitor_err {
+	($self: ident, $err: expr, $entry: expr, $action_type: path, $resend_channel_ready: expr, OPTIONALLY_RESEND_FUNDING_LOCKED) => {
+		handle_monitor_err!($self, $err, $entry, $action_type, false, false, $resend_channel_ready, Vec::new(), Vec::new(), Vec::new())
 	};
-	($self: ident, $err: expr, $channel_state: expr, $entry: expr, $action_type: path, $resend_raa: expr, $resend_commitment: expr) => {
```
Would have preferred to place this cleanup in the final commit, but keeping them in the first commit messed up the macro. I hope that's ok.

ViktorTigerstrom commented:
Realized I need to go through this closer another time, as the current consistency guarantees ensures that the channel is guaranteed to exist in the `by_id` map if also existing in the `short_to_chan_info` map while the channel_state lock is taken. Just need to ensure that we aren't unwrapping the result from getting the channel in the `by_id` map when using an id from the `short_to_chan_info` map, if the channel_state map isn't taken while the id is retrieved from `short_to_chan_info` map.

While open the pull request when I've checked this further.

ViktorTigerstrom commented:
Added a fixup commit that ensures that the consistency guarantees that any `channel_id` in the `short_to_chan_info` map is guaranteed exist as a key in the `channel_state::by_id` map.

Also rebased this on main, now that the blocking PR has been merged.

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -749,6 +743,28 @@ pub struct ChannelManager<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref,
+	/// SCIDs being added once the funding transaction is confirmed at the channel's required
+	/// confirmation depth.
+	///
+	/// Consistancy guarantees with the `channel_state::by_id` map, guarantees that any
```
Ok sure I'll change to that instead :)! I did it this way mostly just to make this PR not change the current consistency guarantees, but if we're ok with changing that, I see no reason why we can't do what you suggest.

ViktorTigerstrom replied:
Fixed with [919ec65](https://github.com/lightningdevkit/rust-lightning/pull/1639/commits/919ec65789faff7a3f9a740cd1951a5ea5ae1a0a)

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1256,9 +1272,12 @@ macro_rules! handle_error {
 macro_rules! update_maps_on_chan_removal {
-	($self: expr, $short_to_chan_info: expr, $channel: expr) => {
+	($self: expr, $channel: expr) => {
+		// The lockorder for id_to_peer is prior to short_to_chan_info.
```
Fixed with [1891450](https://github.com/lightningdevkit/rust-lightning/pull/1639/commits/1891450656ac35f89f65e0324cc94e28897dd6ba)

ViktorTigerstrom commented:
Thanks for the review @TheBlueMatt! I've addressed your feedback with the latest pushed commits. Unfortunately I had to rebase to drop the patch I made before, but I rebased on the same HEAD as before.

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3097,9 +3121,8 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-					let forward_chan_id = match channel_state.short_to_chan_info.get(&short_chan_id) {
-						Some((_cp_id, chan_id)) => chan_id.clone(),
-						None => {
+					macro_rules! fail_pending_forwards {
```
Not super happy about this macro, and would prefer to break it out into a separate function due to the added complexity. But I figured `process_pending_htlc_forwards` was probably kept large with macros for efficacy reasons, so opted for the new macro instead. Let me know if my assumptions are wrong, and you'd prefer to break this out :)

dunxen replied:
I'm not sure here, I just know that nested macros are not fun to debug :see_no_evil:

ViktorTigerstrom replied:
Yeah definitely agree 😬...

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -4098,7 +4129,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			let mut valid_mpp = true;
 			for htlc in sources.iter() {
-				if let None = channel_state.as_ref().unwrap().short_to_chan_info.get(&htlc.prev_hop.short_channel_id) {
+				if let None = self.short_to_chan_info.read().unwrap().get(&htlc.prev_hop.short_channel_id) {
```
Oh yeah good catch! Fixed with 1f01103

ViktorTigerstrom commented:
Thanks! Addressed the latest feedback with 1f01103 :)

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5127,7 +5168,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		let mut channel_state_lock = self.channel_state.lock().unwrap();
 		let channel_state = &mut *channel_state_lock;
-		let chan_id = match channel_state.short_to_chan_info.get(&msg.contents.short_channel_id) {
+		let chan_id = match self.short_to_chan_info.read().unwrap().get(&msg.contents.short_channel_id) {
```
Fixed with 41d275d

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1256,9 +1266,14 @@ macro_rules! handle_error {
 
 macro_rules! update_maps_on_chan_removal {
-	($self: expr, $short_to_chan_info: expr, $channel: expr) => {
+	($self: expr, $channel: expr) => {
```
Oh, thanks for letting me know! Fixed with 91bae8d

dunxen reviewed (commented):
Generally looks good to me. I'll try to give it another pass today, but not 100% confident with pending forwards yet :)

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3097,9 +3123,8 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-					let forward_chan_id = match channel_state.short_to_chan_info.get(&short_chan_id) {
-						Some((_cp_id, chan_id)) => chan_id.clone(),
-						None => {
+					macro_rules! fail_pending_forwards {
+						() => {
```
Think you can get rid of the macro with something like:
```suggestion
					let forward_chan_id_opt = match self.short_to_chan_info.read().unwrap().get(&short_chan_id)
						.map(|(_cp_id, chan_id)| chan_id.clone());
					if forward_chan_id_opt.is_none() || !channel_state.by_id.contains_key(forward_chan_id_opt.unwrap()) {

```
but I haven't actually completed this patch myself

ViktorTigerstrom replied:
Oh, thanks for the suggestion :)! 

That change would indeed work when only taking this PR into account. Sadly though, that change would not help when taking #1507 into consideration as well, which I'd like this PR to prepare for. 1507 is quite huge, so I'd like to do as many changes as possible within this PR, to make 1507 as reviewable as possible.

In 1507, we will lock the new `channel lock` using in this scope by using the info retrieved from `short_to_chan_info` (The `counterparty_node_id` and the `channel_id`). Therefore we wouldn't be able to check that `short_to_chan_info` both contained a value for the `scid` and check that the new `channel lock` contains the channel for the unwrapped `short_to_chan_info` info in the same `if` clause, unless we'd take the new `channel lock` temporarily just for the contains check. Though by taking the `channel lock` temporarily, we would still need the `Vaccant` check below for the `channel lock `, hence requiring the second call for the code in `fail_pending_forwards` :/

valentinewallace replied:
Ah okay, worth a shot!

valentinewallace commented:
I think the fixups may be out of order? Correct me if I'm wrong, but it helps me when a fixup is next to the commit it's fixing up

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2440,7 +2464,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			}
 
-			let id = match channel_lock.short_to_chan_info.get(&path.first().unwrap().short_channel_id) {
+			let id = match self.short_to_chan_info.read().unwrap().get(&path.first().unwrap().short_channel_id) {
```
Oh yes, good catch! Thanks :)

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5579,14 +5620,14 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 	pub fn get_phantom_scid(&self) -> u64 {
-		let mut channel_state = self.channel_state.lock().unwrap();
+		let short_to_chan_info = self.short_to_chan_info.read().unwrap();
 		let best_block = self.best_block.read().unwrap();
```
Thanks :)

ViktorTigerstrom commented:
Thank you for the reviews @valentinewallace and @dunxen :)!

Sorry for taking so long to address the feedback, I've been traveling.
 
> I think the fixups may be out of order? Correct me if I'm wrong, but it helps me when a fixup is next to the commit it's fixing up

Oh yes, you are correct. Thanks for informing me about that it helps in the review process to move up the fixup commits to the correct order. I was incorrectly under the impression that it was less confusing to not move them, but now I know better :). I'll move them up from now on.

ViktorTigerstrom commented:
Rebased on upstream without changes except fixing minor merge conflicts.

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2246,7 +2263,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 				if let Some((err, code, chan_update)) = loop {
 					let mut channel_state = self.channel_state.lock().unwrap();
-					let id_option = channel_state.short_to_chan_info.get(&short_channel_id).cloned();
+					let id_option = self.short_to_chan_info.read().unwrap().get(&short_channel_id).cloned();
```
Yeah sure :)! Sadly we're already holding the `channel_state` when taking the `short_to_chan_info` lock in multiple other places, but might as well fix that here.

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1254,27 +1264,33 @@ macro_rules! handle_error {
-		$short_to_chan_info.remove(&$channel.outbound_scid_alias());
+	($self: expr, $channel: expr) => {
+		{
+			let mut id_to_peer = $self.id_to_peer.lock().unwrap();
```
Good question 😅
Fixed with: e47fc06

ViktorTigerstrom replied:
Yeah I moved this change to the follow-up PR #1795. It's getting a bit unnecessary with a separate PR for the follow-ups now that I'm already doing some small fixes in this PR hah, so I will close that PR and bring the fixups into this PR.

valentinewallace replied:
Understood, since this is basically mergeable I support keeping this fix in followup! But up to you

ViktorTigerstrom replied:
Sorry missed this until I had already pushed the followups with PR yesterday. Hope that's ok :)

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2428,6 +2452,11 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		let err: Result<(), _> = loop {
 			let mut channel_lock = self.channel_state.lock().unwrap();
 
+			let id = match self.short_to_chan_info.read().unwrap().get(&path.first().unwrap().short_channel_id) {
```
e2a0fc8

ViktorTigerstrom replied:
Unfortunately we can't remove the lock order requirement for `short_to_chan_info` to be locked after the `channel_state` lock with this PR, as it's still required in multiple other places in the code, but with #1507 we'll be able to remove that lock order however :).
I think that this order change here and above mainly a preparation for that PR, and is ok here since we only take the `short_to_chan_info`  lock temporarily and drop it before taking the `channel_state` lock.

I agree though that this is a bit confusing, and if you'd like, I could remove this temporary order change.

ViktorTigerstrom replied:
Yes, agree. Once #1772 is merged as well, I'll look into if it makes sense to already change the lock order so that `short_to_chan_info` and `id_to_peer` is before the `channel_state`, as it would be nice to make #1507 less complicated.

ViktorTigerstrom commented:
Addressed the latest feedback, and restricted the access for the `short_to_chain_info` field!

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -5907,14 +5943,15 @@ where
 							// enforce option_scid_alias then), and if the funding tx is ever
 							// un-confirmed we force-close the channel, ensuring short_to_chan_info
 							// is always consistent.
+							let mut short_to_chan_info = self.short_to_chan_info.write().unwrap();
```
Still confused here by the what is the correct lock order between `short_to_chan_info` and `channel_state` one L5895

ViktorTigerstrom replied:
See https://github.com/lightningdevkit/rust-lightning/pull/1639#discussion_r973794535

ariard reviewed (commented):
Looks good, first parse

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2426,8 +2447,12 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		let err: Result<(), _> = loop {
-			let mut channel_lock = self.channel_state.lock().unwrap();
+			let id = match self.short_to_chan_info.read().unwrap().get(&path.first().unwrap().short_channel_id) {
+				None => return Err(APIError::ChannelUnavailable{err: "No channel available with first hop!".to_owned()}),
```
Ah ok, I didn't know this, so thanks for informing me! Is it mainly the error message that we'd like to change, or would we like to create a new `APIError` variant as well?

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2506,7 +2526,9 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 				}
-			} else { unreachable!(); }
+			} else {
+				return Err(APIError::ChannelUnavailable{err: "No channel available with first hop!".to_owned()});
```
Hmm I'm not sure if you mean that we'd like to change the error message here to indicate to the user the lack between  the `short_to_chan_info` and the `by_id` maps, or if you mean that we should just add a code comment here to indicate it. I don't think we should change the message to surface the reason for the problem to the user, as neither the `short_to_chan_info` nor the `by_id` map are public to the user, and it would therefore make no sense to refer to them in the message.

I'll add a code comment here to note why this was triggered :)

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3095,9 +3117,8 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-					let forward_chan_id = match channel_state.short_to_chan_info.get(&short_chan_id) {
-						Some((_cp_id, chan_id)) => chan_id.clone(),
-						None => {
+					macro_rules! fail_pending_forwards {
```
Ah good point! Perhaps we'd like to name the macro to something else than `fail_pending_forwards` then? I'll rename it to `forwarding_channel_not_found` temporarily, but if you have a better suggestion let me know :)

ViktorTigerstrom commented:
Thanks a lot for the review @ariard! Addressed some of your feedback with my latest commits, and left comments for the rest.

ViktorTigerstrom commented:
Rebased on main to fix merge conflicts after #1656 was merged.
Also added 2 new commits:
* [181a959](https://github.com/lightningdevkit/rust-lightning/pull/1639/commits/181a9594cbf128c3b4a6faded9ebbb00d2e00574) To adhere the lock order docs for `short_to_chan_info` to the new format set in #1656. 
* [cd33217](https://github.com/lightningdevkit/rust-lightning/pull/1639/commits/cd3321723027fa520d7df046d7c6ad89f32ae5ce) To adapt this PR to the new type of the `channel_state` variable in `claim_funds`. The "commits" checks are now expected to fail, until this fixup commit has been squashed.

Let me know if you'd like me to squash the fixups :)!

ViktorTigerstrom commented:
No worries, I'll split the follow-up commits into a separate PR and fix the merge conflicts here

ViktorTigerstrom commented:
Fixed the merge conflicts and removed the follow-up PRs that could be removed. Noticed that the follow-up changes I could remove was very tiny though, and might not make sense to have in a separate PR.... I'll push them in a separate PR very soon, but if you think it makes more sense to close that PR and include them in this PR, let me know and I'll include them here instead.

ViktorTigerstrom commented:
Pushed follow-ups in #1795

dunxen commented:
From another quick review I believe this looks good to go after squash.

ViktorTigerstrom commented:
Thanks @dunxen! 

Reviewers, would it be ok if I squash the commits :)?

ViktorTigerstrom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1422,44 +1440,45 @@ macro_rules! handle_monitor_update_res {
 		// Note that we may send a `channel_ready` multiple times for a channel if we reconnect, so
 		// we allow collisions, but we shouldn't ever be updating the channel ID pointed to.
-		let outbound_alias_insert = $short_to_chan_info.insert($channel.outbound_scid_alias(), ($channel.get_counterparty_node_id(), $channel.channel_id()));
+		let mut short_to_chan_info = $self.short_to_chan_info.write().unwrap();
```
Fixed with the latest push. I didn't push in a separate fixup commit, hope that's ok!

ViktorTigerstrom commented:
Squashed and addressed latest feedback :)!

dunxen commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -791,6 +787,22 @@ pub struct ChannelManager<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref>
+	/// guarantees are made about the existence of a peer with the `counterparty_node_id` nor a
+	/// channel with the `channel_id` in our other maps.
+	///
+	/// See `ChannelManager` struct-level documentation for lock order requirements.
```
nit: Making `ChannelManager` a doc link would be helpful to get straight to the struct-level docs.

valentinewallace replied:
Ah I thought it was the case that private fields don't appear on the docs website therefore links wouldn't be clickable anywhere?

dunxen replied:
Oh whoops. Didn't catch this was private :)

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2285,7 +2304,14 @@ impl<M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelManager<M, T, K, F
 					};
 					let chan_update_opt = if let Some(forwarding_id) = forwarding_id_opt {
-						let chan = channel_state.by_id.get_mut(&forwarding_id).unwrap();
+						let chan = match channel_state.by_id.get_mut(&forwarding_id){
```
nit:
```suggestion
						let chan = match channel_state.by_id.get_mut(&forwarding_id) {
```

valentinewallace reviewed (approved):
LGTM, feel free to take or leave nits

ViktorTigerstrom commented:
Thanks @dunxen & @valentinewallace.
Will address the nits and move the follow-ups from #1795 into this PR

ViktorTigerstrom commented:
Addressed the latest feedback, and also brought in the followups from #1795 into this PR (ie. scope fix for `update_maps_on_chan_removal`, as well as locking `short_to_chan_info` temporarily before locking `channel_state` in 2 occasions as a preparation for moving the lock_order for `short_to_chain_info` to before the `channel_state` lock).

I fixed up these changes into the corresponding commit already as they were minor. I hope that's ok :)

valentinewallace commented:
LGTM after rebase

ViktorTigerstrom commented:
Rebased on main. The only changes were merge conflict fixes 🚀!

ViktorTigerstrom commented:
Removed the `is_usable` check and squashed

ViktorTigerstrom commented:
Sorry, reverted once more to _just_ remove the `is_usable` check and not the existence check 😅
